NorthstarDLL/logging.cpp

```diff
@@ -12,257 +11,194 @@
+	(void)ignored;
+}
 
-// clang-format off
-AUTOHOOK(TextMsg, client.dll + 0x198710,
-void,, (BFRead* msg))
-// clang-format on
+void ExternalConsoleSink::flush_()
 {
-	TextMsgPrintType_t msg_dest = (TextMsgPrintType_t)msg->ReadByte();
-
-	char text[256];
-	msg->ReadString(text, sizeof(text));
-
-	if (!Cvar_cl_showtextmsg->GetBool())
-		return;
-
-	switch (msg_dest)
-	{
-	case TextMsgPrintType_t::HUD_PRINTCENTER:
-		pInternalCenterPrint->Print(text);
-		break;
-
-	default:
-		spdlog::warn("Unimplemented TextMsg type {}! printing to console", msg_dest);
-		[[fallthrough]];
-
-	case TextMsgPrintType_t::HUD_PRINTCONSOLE:
-		auto endpos = strlen(text);
-		if (text[endpos - 1] == '\n')
-			text[endpos - 1] = '\0'; // cut off repeated newline
-
-		spdlog::info(text);
-		break;
-	}
+	std::cout << std::flush;
 }
 
-// clang-format off
-AUTOHOOK(ConCommand_echo, engine.dll + 0x123680,
-void,, (const CCommand& arg))
-// clang-format on
+void CustomSink::custom_log(const custom_log_msg& msg)
 {
-	if (arg.ArgC() >= 2)
-		spdlog::info("[echo] {}", arg.ArgS());
+	std::lock_guard<std::mutex> lock(mutex_);
+	custom_sink_it_(msg);
 }
 
-// This needs to be called after hooks are loaded so we can access the command line args
-void CreateLogFiles()
+void InitialiseConsole()
 {
-	if (strstr(GetCommandLineA(), "-disablelogs"))
+	if (AllocConsole() == FALSE)
 	{
-		spdlog::default_logger()->set_level(spdlog::level::off);
+		std::cout << "[*] Failed to create a console window, maybe a console already exists?" << std::endl;
 	}
 	else
 	{
-		try
-		{
-			// todo: might be good to delete logs that are too old
-			time_t time = std::time(nullptr);
-			tm currentTime = *std::localtime(&time);
-			std::stringstream stream;
+		freopen("CONOUT$", "w", stdout);
+		freopen("CONOUT$", "w", stderr);
+	}
 
-			stream << std::put_time(&currentTime, (GetNorthstarPrefix() + "/logs/nslog%Y-%m-%d %H-%M-%S.txt").c_str());
-			spdlog::default_logger()->sinks().push_back(std::make_shared<spdlog::sinks::basic_file_sink_mt>(stream.str(), false));
-			spdlog::flush_on(spdlog::level::info);
-		}
-		catch (...)
+	// this if statement is adapted from r5sdk
+	if (!strstr(GetCommandLineA(), "-noansiclr"))
+	{
+		DWORD dwMode = NULL;
+		HANDLE hOutput = GetStdHandle(STD_OUTPUT_HANDLE);
+
+		GetConsoleMode(hOutput, &dwMode);
+		dwMode |= ENABLE_PROCESSED_OUTPUT | ENABLE_VIRTUAL_TERMINAL_PROCESSING;
+
+		if (!SetConsoleMode(hOutput, dwMode)) // Some editions of Windows have 'VirtualTerminalLevel' disabled by default.
 		{
-			spdlog::error("Failed creating log file!");
-			MessageBoxA(
-				0, "Failed creating log file! Make sure the profile directory is writable.", "Northstar Warning", MB_ICONWARNING | MB_OK);
+			// If 'VirtualTerminalLevel' can't be set, just disable ANSI color, since it wouldnt work anyway.
+			spdlog::warn("could not set VirtualTerminalLevel. Disabling color output");
+			g_bSpdLog_UseAnsiColor = false;
```

Comment:
I believe that this function is called before InitialiseLogging, meaning that g_bSpdLog_UseAnsiColor hasn't been set to true yet, and spdlog::warn might not do anything?